Mark Link Creation Interface as Experimental

marekhrabe opened the pull request:
Sub-PR of #17846, meant to be merged into it, not master. 

It adds experimental flags for the main component and all additions (in form of props) that were added to (already public) `URLInput`.

marekhrabe committed:
mark main component as experimental

marekhrabe committed:
mark new URLInput props as experimental

marekhrabe committed:
add experimental onKeyPress

marekhrabe commented on packages/block-editor/README.md:
```diff
@@ -302,10 +302,6 @@ _Related_
-<a name="LinkControl" href="#LinkControl">#</a> **LinkControl**
-
-Undocumented declaration.
-
```
noting this was removed automattically by a `docs:build` script because all components containing word `unstable` or `experimental` are filtered there

getdave replied:
Ok

marekhrabe commented on packages/block-editor/src/components/url-input/index.js:
```diff
@@ -305,7 +306,7 @@ class URLInput extends Component {
 					placeholder={ placeholder }
 					onKeyDown={ this.onKeyDown }
-					onKeyPress={ this.props.onKeyPress || noop }
+					onKeyPress={ isFunction( __experimentalOnKeyPress ) ? __experimentalOnKeyPress : undefined }
```
Not sure if there is a big performance gain but I thought it might be better not to define any callback rather than passing `noop`.

getdave replied:
So do we need the check at all then? Maybe just pass `__experimentalOnKeyPress` directly and then if it isn't provided then it will be `undefined` right?

marekhrabe replied:
Sure that will do for case where you don't provide anything for the prop. However if you provide something else (like `false`), we'll allow an error if we don't check.

marekhrabe replied:
gone

marekhrabe commented on packages/block-editor/src/components/url-input/index.js:
```diff
@@ -139,8 +139,8 @@ class URLInput extends Component {
 		// Trigger `onKeyDown` event, passing the selected suggestion besides the event class.
-		if ( this.props.onKeyDown ) {
-			this.props.onKeyDown( event, selectedSuggestion );
+		if ( isFunction( this.props.__experimentalOnKeyDown ) ) {
```
I added `isFunction` check before calling it

getdave replied:
I think @talldan specifically advised against this because it wouldn't provide useful error reporting.

marekhrabe replied:
what's your suggested fix? Is `onKeyDown` being completely removed? If so, might be best to just remove it here right away

getdave replied:
Yeh sorry not overly helpful was I!? 😄 

I believe he was saying to just check if `this.props.__experimentalOnKeyDown` was defined. That way if it wasn't it would give a clearer error. This is where I feel we miss prop types.

```js
if ( this.props.__experimentalOnKeyDown  ) {
```

marekhrabe replied:
You were - I got it the first time but I also overheard that this property is possibly going away. That's why I'm asking whether I should bother with fixing or maybe remove it too, as we know that's gonna happen anyway.

marekhrabe replied:
gone

marekhrabe commented on packages/block-editor/src/components/link-control/input-search.js:
```diff
@@ -13,31 +13,31 @@ import { URLInput } from '../';
+	fetchSuggestions,
+	onReset,
+	onKeyDown,
+	onKeyPress,
```
I got the linter screaming at me :D

getdave commented on packages/block-editor/src/components/link-control/input-search.js:
```diff
@@ -13,31 +13,31 @@ import { URLInput } from '../';
-				fetchLinkSuggestions={ fetchSuggestions }
-				handleURLSuggestions={ true }
+				__experimentalRenderSuggestions={ renderSuggestions }
+				__experimentalFetchLinkSuggestions={ fetchSuggestions }
```
I think this should resolve https://github.com/WordPress/gutenberg/pull/17846#discussion_r339006840 for @talldan right?

talldan replied:
Yep, looks good 👍

getdave commented on packages/block-editor/src/components/url-input/index.js:
```diff
@@ -256,7 +256,8 @@ class URLInput extends Component {
 			hasBorder,
-			renderSuggestions,
+			__experimentalRenderSuggestions,
+			__experimentalOnKeyPress,
```
I _think_ @retrofox is going to remove this right Damián?

marekhrabe replied:
gone

getdave commented on packages/block-editor/src/components/url-input/index.js:
```diff
@@ -369,12 +370,12 @@ export default compose(
 		const { getSettings } = select( 'core/block-editor' );
 		return {
-			fetchLinkSuggestions: getSettings().__experimentalFetchLinkSuggestions,
+			__experimentalFetchLinkSuggestions: getSettings().__experimentalFetchLinkSuggestions,
```
This prop `__experimentalFetchLinkSuggestions` was already here before I came along:

https://github.com/WordPress/gutenberg/blob/04e142e9cbd06a45c4ea297ec573d389955c13be/packages/block-editor/src/components/url-input/index.js#L317

I'm not sure we should mark as "experimental"?

talldan replied:
Looking at the code in master, the un-enhanced version of `URLInput` isn't exported, so there was no way to use `fetchLinkSuggestions` previously. The value was always overridden by the wrapping `withSelect`. 

Now that the `withSelect` has an early return, the prop can be used, so I think it should be made experimental until there's more chance to think about if it's needed.

getdave replied:
Ah of course. I see what you mean now. Much appreciated.

getdave commented on playground/src/index.js:
```diff
@@ -16,7 +16,7 @@ import {
 	WritingFlow,
 	ObserveTyping,
-	LinkControl,
+	__experimentalLinkControl,
```
You can nuke this from your PR as we won't be merging _any_ changes to the Playground.

marekhrabe replied:
If it's still in your branch (the one this PR targets), I'd keep it so things keep working. If it was already removed, I can rebase and delete it here too.

getdave replied:
Let's leave it and remove from the main canonical PR.

getdave commented on packages/block-editor/src/components/link-control/input-search.js:
```diff
@@ -13,31 +13,31 @@ import { URLInput } from '../';
 				value={ value }
 				onChange={ onChange }
-				onKeyDown={ ( event, suggestion ) => {
+				__experimentalOnKeyDown={ ( event, suggestion ) => {
```
I think @retrofox will be removing all key handlers but you will need to get him to review this.

marekhrabe replied:
gone

getdave reviewed (changes requested):
Looking good. Thanks for much for doing this. I noticed a few things that may need additional sanity checks or reviews.

marekhrabe committed:
remove key handlers

talldan commented on packages/block-editor/src/components/url-input/index.js:
```diff
@@ -256,7 +251,7 @@ class URLInput extends Component {
 			isFullWidth,
 			hasBorder,
-			renderSuggestions,
+			__experimentalRenderSuggestions,
```
Could potentially use an alias on this and other props, as its really only the public interface that needs to be made experimental:
```
__experimentalRenderSuggestions: renderSuggestions
```

 but I think it's ok either way. I'll leave it up to you.

getdave replied:
Fixed as it makes things easier to read and will avoid merge conflicts.

talldan reviewed (approved):
Thanks for handling this. Looks good to me.

edit: I note there are merge conflicts, so I might need a re-test after rebasing.

getdave commented:
Rebasing here shortly...

getdave committed:
Updates to use alias on experimental props

Addresses https://github.com/WordPress/gutenberg/pull/18110#discussion_r339427180

retrofox reviewed (approved):
LGTM. We can tweak in the `try/unified-link-ui--experimental` if it's needed.

youknowriad commented on packages/block-editor/src/components/index.js:
```diff
@@ -22,7 +22,7 @@ export { default as __experimentalGradientPickerControl } from './gradient-picke
 export { default as InnerBlocks } from './inner-blocks';
 export { default as InspectorAdvancedControls } from './inspector-advanced-controls';
 export { default as InspectorControls } from './inspector-controls';
-export { default as LinkControl } from './link-control';
```
When was this component first introduced?

retrofox replied:
this component is not being used, yet.

retrofox replied:
we are integrating the `<LinkControl />` with the `<NavigationMenuItem />` on this PR: https://github.com/WordPress/gutenberg/pull/18062/

youknowriad replied:
It doesn't mean it was not used by plugins. Did we check at least in the plugins repository? Do you know when we introduce it?

retrofox replied:
This branch was based on [try/unified-link-ui--experimental](https://github.com/WordPress/gutenberg/tree/try/unified-link-ui--experimental), and just merged with this one as well. The `<LinkControl />` component is totally new.

talldan replied:
Yep, as @retrofox mentions this PR wasn't merged to `master` but another branch that still has an open PR. `LinkControl` has never been merged to `master` so all should be fine.
